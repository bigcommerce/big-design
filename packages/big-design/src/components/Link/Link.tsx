import { OpenInNewIcon } from '@bigcommerce/big-design-icons';
import React, { ComponentPropsWithoutRef, forwardRef, memo, Ref } from 'react';

import { MarginProps } from '../../helpers';

import { StyledLink } from './styled';

export interface LinkProps extends ComponentPropsWithoutRef<'a'>, MarginProps {
  readonly children?: React.ReactNode;
  readonly ellipsis?: boolean;
  readonly external?: boolean;
}

interface PrivateProps {
  forwardedRef: Ref<HTMLAnchorElement>;
  isExternal?: boolean;
}

const StyleableLink: React.FC<LinkProps & PrivateProps> = memo((props) => (
  <StyledLink {...props} />
));

export const Link = forwardRef<HTMLAnchorElement, LinkProps>(
  ({ children, external, ...props }, ref) => {
    const isExternal = external && props.target === '_blank';

    return (
      <StyleableLink {...props} forwardedRef={ref} isExternal={isExternal}>
        {isExternal ? <span>{children}</span> : children}
        {isExternal && <OpenInNewIcon size="medium" />}
      </StyleableLink>
    );
  },
);

Link.displayName = 'Link';
