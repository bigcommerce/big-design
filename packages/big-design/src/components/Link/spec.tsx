import { render } from '@testing-library/react';
import React from 'react';

import 'jest-styled-components';

import { Link } from './Link';

test('render link', () => {
  const { container } = render(<Link href="#">Link</Link>);

  expect(container.firstChild).toMatchSnapshot();
});

test('renders with external icon', () => {
  const { container } = render(
    <Link external href="#" target="_blank">
      Link
    </Link>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('forwards styles', () => {
  const { container } = render(
    <Link className="test" href="#" style={{ background: 'red' }}>
      Link
    </Link>,
  );

  expect(container.firstChild).toHaveClass('test');
  expect(container.firstChild).toHaveStyle('background: red');
});

test('render Link with ellipsis', () => {
  const { getByTestId } = render(
    <Link data-testid="link" ellipsis={true}>
      Test with ellipsis
    </Link>,
  );
  const link = getByTestId('link');

  expect(link).toHaveStyle(`
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  `);
});
