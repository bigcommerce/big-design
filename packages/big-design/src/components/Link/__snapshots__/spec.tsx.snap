// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`render link 1`] = `
.c0 {
  -webkit-transition: all 70ms ease-out;
  transition: all 70ms ease-out;
  -webkit-transition-property: color;
  transition-property: color;
  color: #3C64F4;
  cursor: pointer;
  font-size: 1rem;
  font-weight: 400;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active {
  color: #0024A6;
}

.c0:hover:not(:active) {
  color: #0024A6;
}

<a
  class="c0"
  href="#"
>
  Link
</a>
`;

exports[`renders with external icon 1`] = `
.c1 {
  vertical-align: middle;
  height: 1rem;
  width: 1rem;
}

.c0 {
  -webkit-transition: all 70ms ease-out;
  transition: all 70ms ease-out;
  -webkit-transition-property: color;
  transition-property: color;
  color: #3C64F4;
  cursor: pointer;
  font-size: 1rem;
  font-weight: 400;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:active {
  color: #0024A6;
}

.c0:hover:not(:active) {
  color: #0024A6;
}

.c0 svg {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-left: 0.25rem;
}

<a
  class="c0"
  href="#"
  target="_blank"
>
  <span>
    Link
  </span>
  <svg
    aria-hidden="true"
    class="c1"
    fill="currentColor"
    height="24"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 24 24"
    width="24"
  >
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
    <path
      d="M18 19H6c-.55 0-1-.45-1-1V6c0-.55.45-1 1-1h5c.55 0 1-.45 1-1s-.45-1-1-1H5a2 2 0 0 0-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2v-6c0-.55-.45-1-1-1s-1 .45-1 1v5c0 .55-.45 1-1 1M14 4c0 .55.45 1 1 1h2.59l-9.13 9.13a.996.996 0 1 0 1.41 1.41L19 6.41V9c0 .55.45 1 1 1s1-.45 1-1V4c0-.55-.45-1-1-1h-5c-.55 0-1 .45-1 1"
    />
  </svg>
</a>
`;
